File: static/js/models/Tag.js
/**
 * Tag model class.
 */

import { Category } from '/static/js/models/Category.js';
import { Question } from '/static/js/models/Question.js';
import { formToApiJson } from '/static/js/models/helpers.js';

export class Tag {
  /**
   * @param {number}      id
   * @param {string}      title
   * @param {number}      usage
   * @param {number}      categoryId
   * @param {Category|null} [category]
   * @param {Question[]}  [questions]
   * @param {string|null} createdAt
   * @param {string|null} updatedAt
   */
  constructor(id, name, usage, categoryId, category = null, questions = [], createdAt = null, updatedAt = null) {
    this.id = id;
    this.name = name;
    this.title = name;
    this.usage = usage;
    this.categoryId = categoryId;
    this.category = category;
    this.questions = questions;
    this.createdAt = createdAt ? new Date(createdAt) : null;
    this.updatedAt = updatedAt ? new Date(updatedAt) : null;
  }

  /** @param {Object} json @returns {Tag} */
  static fromJSON(json) {
    return new Tag(
      json.id,
      json.name ?? json.title ?? '',
      json.usage,
      json.categoryId,
      json.category ? Category.fromJSON(json.category) : null,
      (json.questions ?? []).map(Question.fromJSON),
      json.createdAt ?? null,
      json.updatedAt ?? null,
    );
  }

  /** @returns {Object} */
  static toModel(form) {
    return formToApiJson(form);
  }
}

File: internal/usecases/dto/tag.go
package dto

import "strconv"

// =====================
// TagForm はタグのフォーム DTO である。
type TagForm struct {
	ID         int64          `json:"id"`
	Name       string         `json:"name"`
	Usage      int            `json:"usage"`
	CategoryID string         `json:"categoryId"`
	CreatedAt  *string        `json:"createdAt,omitempty"`
	UpdatedAt  *string        `json:"updatedAt,omitempty"`
	DeletedAt  *string        `json:"deletedAt,omitempty"`
	Category   *CategoryForm  `json:"category,omitempty"`
	Questions  []QuestionForm `json:"questions,omitempty"`
}

// CategoryIDInt64 は CategoryID 文字列を int64 に変換する。
//
// return:
//   - int64: カテゴリ ID（変換失敗時は -1）
func (tf TagForm) CategoryIDInt64() int64 {
	if val, err := strconv.ParseInt(tf.CategoryID, 10, 64); err == nil {
		return val
	}
	return -1
}
